bootstrap.sh

```diff
@@ -1,117 +1,308 @@
+		echo
+		echo "Please use the SSH keys that you specified in the AWS Management Console to log in to the server."
+		echo "Also, make sure that your Security Group allows inbound connections on 51820/udp, 80/tcp and 443/tcp."
+		echo
+	fi
+}
+
+ssh_keys_non_aws() {
+	echo
+	echo "Would you like to use an existing SSH key?"
+	echo "Press 'n' if you want to generate a new SSH key pair"
+	echo
+	new_ssh_key_pair=
+	until [[ ${new_ssh_key_pair} =~ ^[yYnN].*$ ]]; do
+		[[ -n ${new_ssh_key_pair} ]] && {
+			echo "${new_ssh_key_pair}: invalid selection."
+		}
+		read -r -p "Use existing SSH key? [y/N]: " new_ssh_key_pair
+	done
+	echo "enable_ssh_keygen: true" |tee -a "${CUSTOM_FILE}"
+
+	[[ "${new_ssh_key_pair}" =~ ^[yY].*$ ]] && {
+		# NO checks done for public key....
+		echo
+		read -r -p "Please enter your SSH public key: " ssh_key_pair
+		echo "ssh_public_key: \"${ssh_key_pair}\"" |tee -a "${CUSTOM_FILE}"
+	}
+}
+
+
+# Main
+
+ANSIBLE_WORK_DIR="${HOME}/ansible-easy-vpn"
+GITHUB_REPO="https://github.com/notthebee/ansible-easy-vpn"
+
+#### WHY? It kills the script with this heere... ?????
+## Discard stdin. Needed when running from an one-liner which includes a newline
+#read -r -N 999999 -t 0.001
+
+## Quit on error
+## -- already set in shebang line ????
+#set -e
+
+check_os
+install_packages_for_ansible_and_dependencies
+install_and_setup_homedir_files
+
+
+# Set secure permissions for the Vault file
+SECRET_FILE="${HOME}/ansible-easy-vpn/secret.yml"
+[[ -f "${SECRET_FILE}" ]] && {
+	echo
+	echo "WARNING:"
+	echo "${SECRET_FILE} already exists..."
+	echo
+}
+chmod 600 "${SECRET_FILE}"
+
+
+# Permissions are not critical with the CUSTOM_FILE
+# - secrets are not kept in this file
+CUSTOM_FILE="${HOME}/ansible-easy-vpn/custom.yml"
+[[ -f "${CUSTOM_FILE}" ]] && {
+	echo
+	echo "WARNING:"
+	echo "${CUSTOM_FILE} already exists..."
+	echo
+}
 
-# Check if we're running on an AWS EC2 instance
-set +e
-aws=$(curl -m 5 -s http://169.254.169.254/latest/meta-data/ami-id)
 
-if [[ "$aws" =~ ^ami.*$ ]]; then
-  aws=true
-else
-  aws=false
-fi
-set -e
-touch $HOME/ansible-easy-vpn/custom.yml
 
 
 
 clear
 echo "Welcome to ansible-easy-vpn!"
 echo
 echo "This script is interactive"
-echo "If you prefer to fill in the custom.yml file manually,"
+echo "If you prefer to fill in the ${CUSTOM_FILE} file manually,"
 echo "press [Ctrl+C] to quit this script"
 echo
 echo "Enter your desired UNIX username"
-read -p "Username: " username
-until [[ "$username" =~ ^[a-z0-9]*$ ]]; do
-  echo "Invalid username"
-  echo "Make sure the username only contains lowercase letters and numbers"
-  read -p "Username: " username
+username=
+until [[ ${username} =~ ^[a-z0-9]*$ && -n ${username} ]]; do
+	[[ -n ${username} ]] && echo "Invalid username"
+	echo "Make sure the username only contains lowercase letters and numbers"
+	read -r -p "Username: " username
 done
+# First write to CUSTOM_FILE, old content will be lost
+echo "username: \"${username}\"" |tee "${CUSTOM_FILE}"
```

Comment:
No need – Ansible will warn that multiple values are present when you run the playbook, but will just use the latest value provided. This might result in a 'less clean' file, but I feel that wiping it with every boostrap.sh run is not the best solution

Reply:
So... ```tee -a``` then

Reply:
True – I might return to the `sed` solution sometime in the future.

Reply:
> > Does aws being true or false gets used anywhere else other than the bootstrap.sh ? If not, then those settings could be lost too.
> 
> It gets used here: https://github.com/notthebee/ansible-easy-vpn/blob/main/run.yml#L15
> 
> On a second thought, might be a better idea to just save it to custom.yml lol

Okay, but the setting of ```aws``` to true or false seems no value now.  This is different to ```AWS_EC2```

Reply:
Fair 😄